store.rb

```diff
@@ -35,14 +59,25 @@ def session_id
   def stop()
     @log.warn "stopping the store"
     Process.kill("TERM", Process.pid) unless $EXIT
+    @root_watcher.unsubscribe unless @root_watcher.nil?
+    Thread.kill(@fetch_thread) unless @fetch_thread.nil?
```

Comment:
`Thread.kill` in ruby is horrendously unthreadsafe in Ruby for the same reasons that the analogous behavior was deprecated in java: http://java.sun.com/j2se/1.5.0/docs/guide/misc/threadPrimitiveDeprecation.html .

The suggested way to deal with this is to break into a smaller looped sleep that checks some variable to see whether it should die.  Ruby threading is not great.